psftools/man/psf2zx.1
.\" -*- nroff -*-
.\"
.\" psf2zx.1: psf2zx man page
.\" Copyright (c) 2005, 2007 John Elliott
.\"
.\"
.\"
.\" psftools: Manipulate console fonts in the .PSF format
.\" Copyright (C) 2005, 2007  John Elliott
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH psf2zx 1 "21 June, 2008" "Version 1.0.8" "PSF Tools"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
psf2zx - convert a PSF font to Spectrum format
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.PD 0
.B psf2zx
.RI [ "--tap" ]
.RI [ "--naked" ]
.RI [ "--rom=romfile" ]
.RI [ "--sna=snapfile" ]
.RI [ "--mode=bare" ]
.RI [ "--mode=merge1" ]
.RI [ INPUTFILE 
.RI [ OUTPUTFILE ]]
.P
.PD 1
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
psf2zx converts a PSF font into a font suitable for use on a Sinclair
Spectrum.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.B --p3dos
Sets the output format to Spectrum +3DOS, with a 128-byte header followed
by the font data. This is the default output format.
.TP
.B --tap
Sets the output format to .TAP (Spectrum emulator tape image). The resulting
TAP image will contain a single file called T:font.bin.
.TP
.B --naked
Sets the output format to a simple 768-byte binary file will be
written, with no header or trailer records.
.TP
.B --rom=romfile
Sets the output to be a Spectrum ROM, based on the ROM image in 
.I romfile.
The file written will be a copy of 
.I romfile
but with the font replaced by the result of psf2zx.
.TP
.B --sna=snapfile
.LP
Sets the output to be a .SNA snapshot, based on the snapshot passed as
.I snapfile.
The file written will be a copy of 
.I snapfile
but with the font replaced by the result of psf2zx.
.LP
This depends on the CHARS system variable having been set up so that the 
game outputs using the ROM routines; if the program in the snapshot uses 
some other form of output, it may leave CHARS alone or set it to a random 
value. For example, the font in the game 
.I 2112
cannot be replaced using this method.
.TP
.B --mode=raw
Copies characters 32-127 from the PSF file, without attempting to translate
the pound sign, up arrow or copyright symbol. This is the default conversion
mode if the source file has no Unicode directory.
.TP
.B --mode=merge1
Copies characters from the PSF file, converting the pound sign, up arrow and
copyright symbol. This is the default conversion if the source file has a 
Unicode directory. If this option is selected but the source file has no 
Unicode directory, it will be assumed to be using a variant of the
ISO-8859-1 codepage with the up-arrow at character position 18h (based on
LAT1-08.PSF supplied with Linux).
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
The behaviour of the 'merge1' mode is ugly when applied to non-Unicode PSF 
files.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
zx2psf(1)
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
John Elliott <seasip.webmaster@gmail.com>.
.PP


psftools/NEWS
What's new in psftools-1.0.7:

	* Codepages: The PPCGREEK codepage had the two lower-case sigmas the
	 wrong way round. Various other corrections made to the PPCGREEK and
	 PPC860 codepages.

	* New programs: psf2bbc / bbc2psf to convert to/from a BBC Micro
	 soft font.



psftools/tools/psf2zx.c
/*
    psftools: Manipulate console fonts in the .PSF format
    Copyright (C) 2003  John Elliott

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
#include "cnvshell.h"
#include "psflib.h"
#include "zxflib.h"

/* Convert a PSF to a Spectrum font. Optionally re-map the characters
 * from ISO Latin-1 */

#define MD_BARE   0	/* Just save chars 32-127 as Spectrum font */
#define MD_MERGE1 1	/* Get characters 94, 96 and 127 from their
                         * Latin-1 counterparts */

static int mode     = -1;
static ZXF_FORMAT zxf_format = ZXF_P3DOS;
static char *auxfile = NULL;

static char helpbuf[2048];

/* Program name */
char *cnv_progname = "PSF2ZX";

/* ddash = 1 if option started with a double-dash; else 0 */
/* Return NULL if OK, else error string */
char *cnv_set_option(int ddash, char *variable, char *value)
{
	if (!stricmp(variable, "mode")) 
        {
		if      (!stricmp(value, "raw"  )) mode = MD_BARE;
		else if (!stricmp(value, "merge1")) mode = MD_MERGE1;
		else return "Invalid value for --mode option. Use the --help option for the help screen.";
		return NULL;
	}
	if (!stricmp(variable, "p3dos")) { zxf_format=ZXF_P3DOS; return NULL; }
	if (!stricmp(variable, "tap"))   { zxf_format=ZXF_TAP; return NULL; }
	if (!stricmp(variable, "naked")) { zxf_format=ZXF_NAKED; return NULL; }
	if (!stricmp(variable, "sna"))   
	{ 
		auxfile = malloc(1 + strlen(value));
		if (!auxfile) return "Out of memory";
		strcpy(auxfile, value);
		zxf_format = ZXF_SNA;
	 	return NULL; 
	}
	if (!stricmp(variable, "rom"))   
	{ 
		auxfile = malloc(1 + strlen(value));
		if (!auxfile) return "Out of memory";
		strcpy(auxfile, value);
		zxf_format = ZXF_ROM;
	 	return NULL; 
	}
	if (strlen(variable) > 2000) variable[2000] = 0;
	sprintf(helpbuf, "Unknown option: %s\n", variable);
	return helpbuf;
}


/* Return help string */
char *cnv_help(void)
{
	sprintf(helpbuf, "Syntax: %s psffile zxfont { options }\n\n", 
			cnv_progname);
	strcat (helpbuf, "Options: \n"
		    "        --p3dos          - Output in +3DOS format (default).\n"
		    "        --tap            - Output in TAP format.\n"
		    "        --rom=romfile    - Output a ROM based on romfile.\n"
		    "        --sna=snafile    - Output a snapshot based on snafile.\n"
		    "        --naked          - Output just the font bytes.\n"
                    "        --mode=raw       - Don't translate pound sign,\n"
		    "                           arrow and copyright symbol.\n"
                    "        --mode=merge1    - Translate pound sign, arrow \n"
		    "                           and copyright symbol.\n"
		    "(default is merge1 if the PSF is Unicode, raw if not)\n");
	return helpbuf;
}


/* Plain copy of character bitmap PSF->ZX */
static void copychar(PSF_FILE *src, ZXF_FILE *dst, int from, int to)
{
	psf_dword idx;
	psf_byte *sptr;
	zxf_byte *dptr;
	static psf_byte nil8[8];

/* If the PSF has a Unicode directory, look up the tokens in it */
	if (psf_is_unicode(src))
	{
		sptr = nil8;
		if (!psf_unicode_lookup(src, from, &idx))
			sptr = src->psf_data + 8 *  idx;
/* I don't expect many fonts to have the uparrow. So if that can't be found, 
 * fall back on 0x5E */
		else if (from == 0x2191 && !psf_unicode_lookup(src, 0x5E, &idx))
			sptr = src->psf_data + 8 *  idx;
	}
	else
	{
		/* 0x18 is the most likely place to find an uparrow in a
		 * non-unicode PSF. */
		if (from == 0x2191) from = 0x18;
		sptr = src->psf_data + 8 *  from;
	}
	dptr = dst->zxf_data + 8 * (to - 32);
    
	memcpy(dptr, sptr, 8);
}

/* Do the conversion */
char *cnv_execute(FILE *infile, FILE *outfile)
{	
	int rv, x;
	PSF_FILE psf;
	ZXF_FILE zxf;

	psf_file_new(&psf);
	rv = psf_file_read(&psf, infile);	
	if (rv != PSF_E_OK) return psf_error_string(rv);

	if (psf.psf_height != 8 || psf.psf_width != 8)
	{
		psf_file_delete(&psf);
		return "PSF characters must be 8x8 for a Spectrum font.";
	}
	if (mode == -1)
	{
		if (psf_is_unicode(&psf)) mode = MD_MERGE1;
		else			  mode = MD_BARE;
	}

	zxf_file_new(&zxf);
	if (rv == PSF_E_OK) rv = zxf_file_create(&zxf);	
	if (rv == PSF_E_OK)
	{
            if (mode == MD_BARE)
		 for (x = 32; x <= 127; x++) copychar(&psf, &zxf, x, x);
            else for (x = 32; x <= 127; x++) switch(x)
                {
                case 0x5E: copychar(&psf, &zxf, 0x2191, x); break; 
                case 0x60: copychar(&psf, &zxf, 0xA3, x); break;
                case 0x7F: copychar(&psf, &zxf, 0xA9, x); break;
                default:   copychar(&psf, &zxf, x, x);    break;
                }
	    rv = zxf_file_write(&zxf, outfile, zxf_format, auxfile);
	}
	psf_file_delete(&psf);
	
	if (rv != ZXF_E_OK)
	{
		zxf_file_delete(&zxf);
		return zxf_error_string(rv);
	}
	zxf_file_delete(&zxf);
	if (auxfile) free(auxfile);
	return NULL;
}

